grakn-graph/src/main/java/ai/grakn/graph/internal/RelationTypeImpl.java

```diff
@@ -80,6 +76,13 @@ public Relation addRelation() {
     public RelationType hasRole(RoleType roleType) {
         checkTypeMutation();
         putEdge(roleType, Schema.EdgeLabel.HAS_ROLE);
+
+        //Cache the Role internally
+        cachedHasRoles.get().add(roleType);
```

Comment:
We only call `isPresent` when deleting. That is because when deleting the cache only needs to be updated if if has been loaded. But when we adding something either the cache should not be updated at all (which might be a better approach) or it has to be fully updated so that we can be sure the cache contains everything.